indexing-service/src/main/java/io/druid/indexing/overlord/TaskLockbox.java

```diff
@@ -189,44 +178,52 @@ public int compare(Pair<Task, TaskLock> left, Pair<Task, TaskLock> right)
   }
 
   /**
-   * Acquires a lock on behalf of a task. Blocks until the lock is acquired. Throws an exception if the lock
-   * cannot be acquired.
+   * Acquires a lock on behalf of a task. Blocks until the lock is acquired.
    *
    * @param task task to acquire lock for
    * @param interval interval to lock
    * @return acquired TaskLock
    *
-   * @throws java.lang.InterruptedException if the lock cannot be acquired
+   * @throws java.lang.InterruptedException if the current thread is interrupted
    */
   public TaskLock lock(final Task task, final Interval interval) throws InterruptedException
   {
-    long timeout = lockTimeoutMillis;
-    giant.lock();
+    giant.lockInterruptibly();
     try {
       Optional<TaskLock> taskLock;
       while (!(taskLock = tryLock(task, interval)).isPresent()) {
-        long startTime = System.currentTimeMillis();
-        lockReleaseCondition.await(timeout, TimeUnit.MILLISECONDS);
-        long timeDelta = System.currentTimeMillis() - startTime;
-        if (timeDelta >= timeout) {
-          log.error(
-              "Task [%s] can not acquire lock for interval [%s] within [%s] ms",
-              task.getId(),
-              interval,
-              lockTimeoutMillis
-          );
+        lockReleaseCondition.await();
+      }
+      return taskLock.get();
+    }
+    finally {
+      giant.unlock();
+    }
+  }
 
-          throw new InterruptedException(String.format(
-              "Task [%s] can not acquire lock for interval [%s] within [%s] ms",
-              task.getId(),
-              interval,
-              lockTimeoutMillis
-          ));
-        } else {
-          timeout -= timeDelta;
+  /**
+   * Acquires a lock on behalf of a task, waiting up to the specified wait time if necessary.
+   *
+   * @param task      task to acquire a lock for
+   * @param interval  interval to lock
+   * @param timeoutMs maximum time to wait
+   *
+   * @return acquired lock
+   *
+   * @throws InterruptedException if the current thread is interrupted
+   */
+  public TaskLock lock(final Task task, final Interval interval, long timeoutMs) throws InterruptedException
+  {
+    long nanos = TIME_UNIT.toNanos(timeoutMs);
```

Comment:
Milliseconds are anyway hardcoded (parameter name timeout**Ms**) so IMO there is no point in constant TIME_UNIT, it could be just `TimeUnit.MILLISECONDS.toNanos(timeoutMs)`